### PR TITLE
Notebook protocol go-to-definition support

### PR DESCRIPTION
**Pull request by jasongrout**
This is a work-in-progress PR to add go to definition support on top of the notebook protocol developed in #389 by @tkrabel-db.

Likely each plugin, such as go to definition, will need to be modified to support the notebook syncing messages. We'll try to abstract out general patterns where possible, though.

### CODE REVIEW

**Comment by jasongrout** [comment]
I think this minimally works, and the test passes now.

**Comment by ccordoba12** [comment]
Hey @jasongrout, nice to see you here! I took a quick look at PR #389 and it looks quite good, so I'll try to merge it next week.

**Comment by jasongrout** [comment]
> Hey @jasongrout, nice to see you here! I took a quick look at PR #389 and it looks quite good, so I'll try to merge it next week.

Awesome! @tkrabel-db and I are using python-lsp-server at Databricks for LSP support, and we are excited to contribute things like notebook protocol support that would be helpful for the whole community. As you saw on the other ticket, @tkrabel-db is eager to iterate on this repo to help maintain and improve the dev workflow as well.

**Comment by ccordoba12** [comment]
> As you saw on the other ticket, @tkrabel-db is eager to iterate on this repo to help maintain and improve the dev workflow as well.

Great! That would be really helpful because I don't have much time lately to maintain this project.

**Comment by jasongrout** [comment]
I rebased this PR on top of the develop branch, now that #389 is merged.

**Comment by ccordoba12** [comment]
@jasongrout, this only requires to fix the code style issues reported by our failing workflow. The rest looks good to me.

**Comment by jasongrout** [comment]
I pushed some changes that address the pylint errors. The biggest change is factoring out extracting the cell data (pylint was complaining about too many local variables, as a hint that the complexity was too large).

Since the black formatting PR was merged, of course there are conflicts. My hope is that running black on this pr will be enough to resolve those, though.

**Comment by jasongrout** [comment]
@ccordoba12, @tkrabel-db - I think this is ready for review again.

File: test/test_notebook_document.py [hunk 1]
```diff
@@ -582,3 +582,76 @@ def test_notebook__did_close(
         )
         wait_for_condition(lambda: mock_notify.call_count >= 2)
         assert len(server.workspace.documents) == 0
+
+
+def test_notebook_definition(
+    client_server_pair,
+):  # pylint: disable=redefined-outer-name
```
Comment (ccordoba12) [inline test/test_notebook_document.py]:
```suggestion
def test_notebook_definition(client_server_pair):
```

This shouldn't be necessary after PR #418.

**Review by ccordoba12 — COMMENTED** [review]
@jasongrout, please merge again with develop to get the changes done in PR #418 and see if your PR passes with them.

File: test/test_notebook_document.py (hunk 1 above)
Comment (jasongrout) [reply test/test_notebook_document.py]: @ccordoba12 - done!

**Comment by ccordoba12** [comment]
@jasongrout, please run `black` again on `test_notebook_document.py` to make our style checks pass.

Then I'll merge, unless @tkrabel-db has something else to say.

**Comment by jasongrout** [comment]
> @jasongrout, please run `black` again on `test_notebook_document.py` to make our style checks pass.

Done!

**Review by ccordoba12 — APPROVED** [review]
Thanks @jasongrout!

**Comment by jasongrout** [comment]
@ccordoba12 - we're trying to figure out a plan for deploying the changes in this PR. Do you have a rough idea of a timeline for releasing 1.8.0? If it's going to be a while, would it be possible to do a 1.8.0 prerelease? What can we do to help with the 1.8.0 release?

**Comment by ccordoba12** [comment]
> What can we do to help with the 1.8.0 release?

There are some [minor issues](https://github.com/python-lsp/python-lsp-server/milestone/22) that I'd like to address for 1.8.0, so if you could give me a hand with them, I'd really appreciate it. They have to do with preventing simple errors here and there.

**Comment by smacke** [comment]
> There are some [minor issues](https://github.com/python-lsp/python-lsp-server/milestone/22) that I'd like to address for 1.8.0, so if you could give me a hand with them, I'd really appreciate it.

I actually didn't see this before I started going through, but I think I have PRs open now that should address all of these:

- https://github.com/python-lsp/python-lsp-jsonrpc/pull/20 for allowing method dispatcher to respond with errors to the json rpc library
- https://github.com/python-lsp/python-lsp-server/pull/432 to use that to respond with an invalid request for any requests that come in post-shutdown
- https://github.com/python-lsp/python-lsp-server/pull/433 to handle syntax errors with empty text due to invalid encodings
- https://github.com/python-lsp/python-lsp-server/pull/434 to deal with the issue on windows that doesn't like an empty cwd for Popen kwargs

Since the first one requires a change on the json rpc side, it probably needs a version bump in the requirements there as well -- let me know if I can help with that.

**Comment by ccordoba12** [comment]
Thanks a lot @smacke for your help with this! I'll review your PRs tomorrow.

> Since the first one requires a change on the json rpc side, it probably needs a version bump in the requirements there as well -- let me know if I can help with that.

No worries, I'll take care of that when doing the release.

**Comment by smacke** [comment]
> Thanks a lot @smacke for your help with this! I'll review your PRs tomorrow.

No problem and awesome! Thanks a bunch Carlos.

**Comment by ccordoba12** [comment]
I just released 1.8.0. Thanks guys for all your help!

**Pull request merged**